sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -342,15 +342,35 @@ private[sql] class ParquetRelation2(
     // Schema of the whole table, including partition columns.
     var schema: StructType = _
 
+    // Cached leaf statuses
+    var localCachedLeafStatuses: Set[FileStatus] = _
+
+    var lastRefreshTime: Long = 0
+
+    // Cached leaves
+    var cachedLeaves: Array[FileStatus] = Array()
+
     /**
      * Refreshes `FileStatus`es, footers, partition spec, and table schema.
      */
     def refresh(): Unit = {
+      // Check if cachedLeafStatuses is changed or not
+      val leafStatusesChanged = localCachedLeafStatuses != cachedLeafStatuses()
+
       // Lists `FileStatus`es of all leaf nodes (files) under all base directories.
-      val leaves = cachedLeafStatuses().filter { f =>
-        isSummaryFile(f.getPath) ||
-          !(f.getPath.getName.startsWith("_") || f.getPath.getName.startsWith("."))
-      }.toArray
+      val leaves = if (leafStatusesChanged) {
+        localCachedLeafStatuses = cachedLeafStatuses()
```

Comment:
Should we do a deep copy here? Currently it's OK because `cachedLeafStatuses()` always returns a new instance of `Set[FileStatus]`, but it's possible that we use a mutable set object in the future. In that case, the `!=` predicate above will always be true.